File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        '''
        Time: O(n)
        Space: O(1)
        '''
        res = 0
        count = defaultdict(int)
        left = 0

        for right, char in enumerate(s):
            count[char] += 1
            while count[char] > 1:
                count[s[left]] -= 1
                left += 1
            res = max(res, right - left + 1)

        return res

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0, left = 0;
        unordered_map<char, int> count;

        for (int right = 0; right < s.size(); right++) {
            char ch = s[right];
            count[ch]++;
            while (count[ch] > 1) {
                count[s[left]]--;
                left++;
            }
            res = max(res, right - left + 1);
        }

        return res;
    }
};